### ISSUE TITLE
DataGrid Cancel Edit Crashes with a template column binding to Run.Text

### ISSUE THREAD

**Issue by kmzeise** [dotnet/wpf#8816]
### Description

In a data grid setup with 
```<DataGridTextColumn />
<DataGridTemplateColumn Header="Crash" IsReadOnly="True">
  <DataGridTemplateColumn.CellTemplate>
    <DataTemplate>
      <TextBlock>
        <Run Text="{Binding SomeValue}" />
      </TextBlock>
    </DataTemplate>
</DataGridTemplateColumn>
```

This crashes when I edit the text column and then cancel it with ESC.

The Exception is thrown in `VisualTreeUtils.EnsureVisual`, with text `'{0}' is not a Visual or Visual3D.`
 

### Reproduction Steps

Create a WPF application for .NET Framework 4.8
Put a Datagrid on the window, and add the template column shown above

### Expected behavior

Does not throw an exception

### Actual behavior

Throws an exception

### Known Workarounds

If you add to the binding a `Mode=OneWay` specification, the crash does not occur

### Impact

Need to find the workaround and apply it in code, adding unnecessary complexity.

### Configuration

Windows 10, fully updated February 2023, .NET Framework 4.8.1

### Other information

When restoring information, the code seems to go through the bindings, and assumes that all are bound to a Visual. But a `Run` is not a visual.

**Comment by miloush** [comment]
Please provide a repro project.

**Comment by kmzeise** [comment]
[DataGridBindingBug.zip](https://github.com/dotnet/wpf/files/14301649/DataGridBindingBug.zip)
 
Here you go

**Comment by miloush** [comment]
Indeed, assuming the binding target is a visual looks like a bug in `DataGridHelper.UpdateTarget` and also `DataGridHelper.BindingExpressionBelongsToElement`, which possibly causes other issues (binding expression leak, stale validation). These happen to be the only two methods calling `VisualTreeHelper.IsAncestorOf`, so theoretically `ContentElement`s could be taken care of there, but it probably would be more correct to add it to the two methods individually.

The question, however, is whether the assumption that the element must be in the tree is correct. Why not update the binding target regardless of whether it is in the datagrid or not?

**Comment by kmzeise** [comment]
I wonder whether the check is there because of templates that could be bound to something outside the grid as well?

I found it surprising that the Datagrid was saving old state for columns marked as read only. Is that extra work done because distinguishing between read write and read only columns would be too expensive?

**Comment by miloush** [comment]
Read-only cell basically means that you don't get the editing "view" and behavior for them. If you put a textbox there that is on you (and there might be reasons to do so). 

Sure you can have elements bound to stuff outside of datagrid, but I don't see much harm in updating them if the cell is edited. On the other side, you can have elements in the adorner layer that are not part of the cell visual tree and these presumably would currently not be updated either.
